### PR TITLE
Support empty tensor concats in Concat op

### LINKED ISSUE #681: failed to concat a tensor with an empty tensor
**Describe the bug**
When concat a tensor with an empty tensor, namely np.array([]), onnxruntime will throw floating point exception.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 0.2.1
- Python version: 3.6
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**To Reproduce**
Here is the code to generate a sample model: 
``` python
import onnx
from onnx import helper, numpy_helper
from onnx import AttributeProto, TensorProto, GraphProto
import numpy as np
import onnxruntime

# The protobuf definition can be found here:
# https://github.com/onnx/onnx/blob/master/onnx/onnx.proto


# Create one input (ValueInfoProto)
A = helper.make_tensor_value_info('A', TensorProto.FLOAT, [2])
B = helper.make_tensor_value_info('B', TensorProto.FLOAT, [0])

# Create one output (ValueInfoProto)
Y = helper.make_tensor_value_info('Y', TensorProto.FLOAT, [2])

# Create a node (NodeProto)
node_def = helper.make_node(
    'Concat', # node name
    inputs=['B', 'A'], # inputs
    outputs=['Y'], # outputs
    axis=0,
)

initializers = [
    onnx.helper.make_tensor(
     name='A',
     data_type=onnx.TensorProto.FLOAT,
     dims=[2],
     vals=[1,2]
    ),
    numpy_helper.from_array(
        np.array([], dtype=np.float32),
        name='B'
    )
]

# Create the graph (GraphProto)
graph_def = helper.make_graph(
    [node_def],
    "test-model",
    [A, B],
    [Y],
    initializers
)

# Create the model (ModelProto)
model_def = helper.make_model(graph_def,
                              producer_name='onnx-example')

print('The ir_version in model: {}\n'.format(model_def.ir_version))
print('The producer_name in model: {}\n'.format(model_def.producer_name))
print('The graph in model:\n{}'.format(model_def.graph))
onnx.checker.check_model(model_def)
print('The model is checked!')
onnx.save(model_def, "model.onnx")

sess = onnxruntime.InferenceSession("model.onnx")
lotus_result = sess.run(['Y'], {})
print(lotus_result)
```

**Expected behavior**
In tensorflow and other framework, this kind of concat will do nothing but just pass the non-empty tensor to output.

### PR DESCRIPTION
**Pull request by hariharans29**
Resolves #681

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 1]
```diff
@@ -17,57 +17,67 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
...
+      ORT_RETURN_IF_NOT(inputs_n_dims[axis_index] == inputs_0_dims[axis_index],
+		                "Non concat axis dimensions must match: Axis ", axis_index, " has mismatched dimensions of ", inputs_n_dims[axis_index], " and ", inputs_0_dims[axis_index]);
     }
   }
 
   // Calculate the size of the concatenated axis, and verify all other dimensions match
   size_t concat_axis_size = 0;
   for (int index = 0; index < input_count; index++) {
     tensor_pointer = ctx->Input<Tensor>(index);
-    if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: Redundant - checks on lines 18 and 29 already cover these

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 2]
```diff
@@ -17,57 +17,67 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
...
   p.output_axis_pitch = 1;
   for (auto i = int64_t(dims.size()); i-- > axis;)
     p.output_axis_pitch *= dims[i];
 
-  auto& concat_result = *ctx->Output(0, outputShape);
-  p.output_tensor = &concat_result;
-
   for (int input_index = 0; input_index < input_count; input_index++) {
     const Tensor* data_n_ptr = ctx->Input<Tensor>(input_index);
-    ORT_ENFORCE(data_n_ptr != nullptr);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: Redundant - checks on lines 18 and 29 already cover these

**Comment by hariharans29** [comment]
CC: @snnn @pranavsharma @skottmckay

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 3]
```diff
@@ -17,59 +17,82 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
...
     tensor_pointer = ctx->Input<Tensor>(index);
     if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
-    auto& data_n = *tensor_pointer;
-    // Ensure all the other axes match
-    auto dimension_count = inputs_0.Shape().NumDimensions();
-    for (int axis_index = 0; axis_index < dimension_count; axis_index++) {
+    auto& inputs_n = *tensor_pointer;
+    const auto& inputs_n_dims = inputs_n.Shape().GetDims();
+    const size_t inputs_n_rank = inputs_n_dims.size();
+    ORT_ENFORCE(inputs_n_rank == inputs_0_rank, "Ranks of input data are different, cannot concatenate them");
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: please print the ranks for better debugging

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 4]
```diff
@@ -17,59 +17,82 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
...
+    // Ensure all the other (non-concat) axes match
+    for (int axis_index = 0; axis_index < inputs_0_rank; axis_index++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: prefer using ++axis_index as a habit, although it doesn't matter here so much

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 5]
```diff
@@ -17,59 +17,82 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
...
+  size_t num_elements = 1; // cache size of the first input along the way
+  for (int dimension_index = 0; dimension_index < inputs_0_rank; dimension_index++) {
+    dims[dimension_index] = inputs_0_dims[dimension_index];
+    num_elements *= inputs_0_dims[dimension_index];
+  }
+  tensor_num_elements[0] = num_elements;
   dims[axis] = concat_axis_size;
   TensorShape outputShape(dims);
+ 
+  auto& concat_result = *ctx->Output(0, outputShape);
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: name outputShape doesn't follow conventions.

File: onnxruntime/core/providers/cpu/tensor/concat.cc [hunk 6]
```diff
@@ -17,59 +17,82 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
...
     auto& data_n = *data_n_ptr;
 
     ORT_RETURN_IF_NOT(data_n.DataType() == concat_result.DataType());
 
     // The input_axis_pitch is the number of elements to add to move to the next split axis in the input
     int64_t input_axis_pitch = 1;
-    for (int i = int(data_n.Shape().NumDimensions()); i-- > axis;)
-      input_axis_pitch *= data_n.Shape()[i];
+    const auto& data_dims = data_n.Shape().GetDims();
+    for (int i = int(inputs_0_rank); i-- > axis;)
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/concat.cc]: use c++ style cast: static_cast<int>()

File: onnxruntime/core/providers/cpu/tensor/concat.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/concat.cc]: This is from the earlier implementation as is - but will fix

File: onnxruntime/core/providers/cpu/tensor/concat.cc (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/concat.cc]: Again - just stuck to the style of the earlier implementation - will fix

**Pull request merged**